atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java

```diff
@@ -652,6 +634,94 @@ public String toString() {
         }
     }
 
+    private Map<Cell, Value> fetchAllCellsForRows(final InetSocketAddress host,
+            final TableReference tableRef,
+            final List<byte[]> rows,
+            final long startTs) throws Exception {
+
+        final ListMultimap<ByteBuffer, ColumnOrSuperColumn> result = LinkedListMultimap.create();
+
+        List<KeyPredicate> query = rows.stream()
+                .map(row -> keyPredicate(ByteBuffer.wrap(row), allPredicateWithLimit(config.fetchReadLimitPerRow())))
+                .collect(Collectors.toList());
+
+        while (!query.isEmpty()) {
+            ListMultimap<ByteBuffer, ColumnOrSuperColumn> partialResult = KeyedStream.stream(fetchCellsForKeyPredicates(host, tableRef, query, startTs))
+                    .filter(cells -> !cells.isEmpty())
+                    .flatMap(Collection::stream)
+                    .collectToMultimap(LinkedListMultimap::create);
+
+            result.putAll(partialResult);
+
+            query = KeyedStream.stream(Multimaps.asMap(partialResult))
+                    .map((row, cells) -> keyPredicate(row, getNextLexicographicalSlicePredicate(cells)))
+                    .values()
+                    .collect(Collectors.toList());
+        }
+
+        ValueExtractor extractor = new ValueExtractor(metricsManager, Maps.newHashMapWithExpectedSize(result.size()));
+        extractor.extractResults(Multimaps.asMap(result), startTs, ColumnSelection.all());
+        return extractor.asMap();
+    }
+
+    private static KeyPredicate keyPredicate(ByteBuffer row, SlicePredicate predicate) {
+        return new KeyPredicate().setKey(row).setPredicate(predicate);
+    }
+
+    private static SlicePredicate allPredicateWithLimit(int limit) {
+        return SlicePredicates.create(Range.ALL, Limit.of(limit));
+    }
+
+    private Map<ByteBuffer, List<ColumnOrSuperColumn>> fetchCellsForKeyPredicates(final InetSocketAddress host,
```

Comment:
as above, I'd prefer `get` instead of `fetch` because we are getting the values. Strictly speaking it should be `getForCellsForKeyPredicates` but I find that a bit clumsy, so would be ok with just changing the first word.